File: python/src/data_platform_lab/orchestration/runner.py
"""Orchestration runner — sequential pipeline execution engine."""

from __future__ import annotations

import datetime
import logging
import time
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any


logger = logging.getLogger(__name__)


@dataclass
class StepDefinition:
    """A registered step in the pipeline."""

    name: str
    fn: Callable[[dict[str, Any]], Any]
    retries: int = 0  # max retry attempts (0 = no retries)
    allow_skip: bool = False  # if True, failure is non-fatal


@dataclass
class StepResult:
    """Result of executing a single step."""

    name: str
    status: str  # "success", "failed", "skipped"
    started_at: str  # ISO timestamp
    ended_at: str  # ISO timestamp
    duration_seconds: float
    result: Any = None  # return value from the step function
    error: str | None = None  # error message if failed
    attempts: int = 1  # total attempts (1 = no retries needed)


@dataclass
class PipelineResult:
    """Result of executing the full pipeline."""

    pipeline_name: str
    status: str  # "success", "failed"
    started_at: str
    ended_at: str
    duration_seconds: float
    steps: list[StepResult] = field(default_factory=list)
    steps_passed: int = 0
    steps_failed: int = 0
    steps_skipped: int = 0


class Pipeline:
    """Sequential pipeline runner.

    Usage::

        pipeline = Pipeline("my_etl")
        pipeline.add_step("extract", extract_fn)
        pipeline.add_step("validate", validate_fn, retries=2)
        pipeline.add_step("transform", transform_fn)
        pipeline.add_step("load", load_fn, allow_skip=True)
        pipeline.add_step("report", report_fn)

        result = pipeline.run()
        print(format_result(result))
    """

    def __init__(self, name: str) -> None:
        self._name = name
        self._steps: list[StepDefinition] = []
        self._logger = logging.getLogger(f"{__name__}.{name}")

    @property
    def name(self) -> str:
        return self._name

    @property
    def steps(self) -> list[StepDefinition]:
        return list(self._steps)

    def add_step(
        self,
        name: str,
        fn: Callable[[dict[str, Any]], Any],
        *,
        retries: int = 0,
        allow_skip: bool = False,
    ) -> Pipeline:
        """Register a step. Returns self for chaining."""
        self._steps.append(
            StepDefinition(name=name, fn=fn, retries=retries, allow_skip=allow_skip)
        )
        return self

    def run(self, context: dict[str, Any] | None = None) -> PipelineResult:
        """Execute all registered steps in order.

        Each step receives the shared *context* dict.  A step's return value
        is stored in ``context["step_results"][step_name]``.

        If a step fails and ``allow_skip`` is ``False``, execution stops
        immediately (fail-fast).  If ``allow_skip`` is ``True``, the step is
        marked ``"skipped"`` and execution continues.

        When ``retries > 0``, the step is retried up to that many additional
        times before being considered failed.
        """
        pipeline_start = datetime.datetime.now(datetime.UTC)
        pipeline_clock = time.perf_counter()

        if context is None:
            context = {}
        context["pipeline_name"] = self._name
        context["step_results"] = {}

        step_results: list[StepResult] = []
        pipeline_failed = False

        for step in self._steps:
            step_start = datetime.datetime.now(datetime.UTC)
            step_clock = time.perf_counter()

            last_error: Exception | None = None
            result_value: Any = None
            max_attempts = 1 + step.retries
            attempts = 0
            succeeded = False

            for attempt in range(1, max_attempts + 1):
                attempts = attempt
                try:
                    self._logger.debug(
                        "Running step %r (attempt %d/%d)",
                        step.name,
                        attempt,
                        max_attempts,
                    )
                    result_value = step.fn(context)
                    succeeded = True
                    break
                except Exception as exc:  # noqa: BLE001
                    last_error = exc
                    self._logger.debug(
                        "Step %r attempt %d failed: %s",
                        step.name,
                        attempt,
                        exc,
                    )

            step_end = datetime.datetime.now(datetime.UTC)
            step_duration = time.perf_counter() - step_clock

            if succeeded:
                context["step_results"][step.name] = result_value
                step_results.append(
                    StepResult(
                        name=step.name,
                        status="success",
                        started_at=step_start.isoformat(),
                        ended_at=step_end.isoformat(),
                        duration_seconds=step_duration,
                        result=result_value,
                        attempts=attempts,
                    )
                )
                self._logger.info("Step %r succeeded", step.name)
            else:
                error_msg = str(last_error)
                if step.allow_skip:
                    step_results.append(
                        StepResult(
                            name=step.name,
                            status="skipped",
                            started_at=step_start.isoformat(),
                            ended_at=step_end.isoformat(),
                            duration_seconds=step_duration,
                            error=error_msg,
                            attempts=attempts,
                        )
                    )
                    self._logger.warning(
                        "Step %r failed but is skippable: %s",
                        step.name,
                        error_msg,
                    )
                else:
                    step_results.append(
                        StepResult(
                            name=step.name,
                            status="failed",
                            started_at=step_start.isoformat(),
                            ended_at=step_end.isoformat(),
                            duration_seconds=step_duration,
                            error=error_msg,
                            attempts=attempts,
                        )
                    )
                    self._logger.error(
                        "Step %r failed: %s", step.name, error_msg
                    )
                    pipeline_failed = True
                    break

        pipeline_end = datetime.datetime.now(datetime.UTC)
        pipeline_duration = time.perf_counter() - pipeline_clock

        steps_passed = sum(1 for s in step_results if s.status == "success")
        steps_failed = sum(1 for s in step_results if s.status == "failed")
        steps_skipped = sum(1 for s in step_results if s.status == "skipped")

        return PipelineResult(
            pipeline_name=self._name,
            status="failed" if pipeline_failed else "success",
            started_at=pipeline_start.isoformat(),
            ended_at=pipeline_end.isoformat(),
            duration_seconds=pipeline_duration,
            steps=step_results,
            steps_passed=steps_passed,
            steps_failed=steps_failed,
            steps_skipped=steps_skipped,
        )


def format_result(result: PipelineResult) -> str:
    """Format a :class:`PipelineResult` as a human-readable string."""
    total = len(result.steps)
    lines: list[str] = [
        f"=== Pipeline: {result.pipeline_name} ===",
        f"Status: {result.status}",
        f"Duration: {result.duration_seconds:.2f}s",
        (
            f"Steps: {total} total | {result.steps_passed} passed"
            f" | {result.steps_failed} failed"
            f" | {result.steps_skipped} skipped"
        ),
        "",
    ]

    for step in result.steps:
        if step.status == "success":
            tag = "PASS"
        elif step.status == "skipped":
            tag = "SKIP"
        else:
            tag = "FAIL"

        detail = f"{step.duration_seconds:.2f}s"
        if step.attempts > 1:
            detail += f", {step.attempts} attempts"

        line = f"  [{tag}] {step.name} ({detail})"
        if step.error is not None:
            line += f" \u2014 {step.error}"
        lines.append(line)

    return "\n".join(lines)

File: javascript/src/orchestration/runner.js
/**
 * Sequential pipeline runner — executes ordered steps with retry,
 * skip-on-failure, and structured run summaries.
 */

/**
 * Sequential pipeline runner.
 *
 * @example
 * const pipeline = new Pipeline("my_etl");
 * pipeline.addStep("extract", extractFn);
 * pipeline.addStep("validate", validateFn, { retries: 2 });
 * pipeline.addStep("transform", transformFn);
 * pipeline.addStep("load", loadFn, { allowSkip: true });
 * pipeline.addStep("report", reportFn);
 *
 * const result = await pipeline.run();
 * console.log(formatResult(result));
 */
export class Pipeline {
  /**
   * @param {string} name - Pipeline name
   */
  constructor(name) {
    this._name = name;
    /** @type {Array<{name: string, fn: Function, retries: number, allowSkip: boolean}>} */
    this._steps = [];
  }

  /** @returns {string} */
  get name() {
    return this._name;
  }

  /** @returns {Array<{name: string, fn: Function, retries: number, allowSkip: boolean}>} */
  get steps() {
    return this._steps;
  }

  /**
   * Register a step. Returns this for chaining.
   * @param {string} name
   * @param {Function} fn - async function receiving context object
   * @param {object} [options]
   * @param {number} [options.retries=0] - max retry attempts
   * @param {boolean} [options.allowSkip=false] - if true, failure is non-fatal
   * @returns {Pipeline}
   */
  addStep(name, fn, options = {}) {
    const { retries = 0, allowSkip = false } = options;
    this._steps.push({ name, fn, retries, allowSkip });
    return this;
  }

  /**
   * Execute all registered steps in order.
   * @param {object} [context={}] - shared context passed to every step
   * @returns {Promise<{
   *   pipeline_name: string,
   *   status: "success"|"failed",
   *   started_at: string,
   *   ended_at: string,
   *   duration_seconds: number,
   *   steps: Array<{
   *     name: string,
   *     status: "success"|"failed"|"skipped",
   *     started_at: string,
   *     ended_at: string,
   *     duration_seconds: number,
   *     result: any,
   *     error: string|null,
   *     attempts: number
   *   }>,
   *   steps_passed: number,
   *   steps_failed: number,
   *   steps_skipped: number
   * }>}
   */
  async run(context = {}) {
    const pipelineStartedAt = new Date().toISOString();
    const pipelineStartMs = Date.now();

    context.pipeline_name = this._name;
    context.step_results = {};

    const stepResults = [];
    let pipelineFailed = false;

    for (const step of this._steps) {
      const stepStartedAt = new Date().toISOString();
      const stepStartMs = Date.now();

      let result = undefined;
      let error = null;
      let status = "success";
      let attempts = 0;
      const maxAttempts = 1 + step.retries;

      for (let attempt = 1; attempt <= maxAttempts; attempt++) {
        attempts = attempt;
        try {
          result = await step.fn(context);
          error = null;
          break;
        } catch (err) {
          error = err.message ?? String(err);
          result = undefined;
        }
      }

      const stepEndedAt = new Date().toISOString();
      const stepDuration = (Date.now() - stepStartMs) / 1000;

      if (error !== null) {
        if (step.allowSkip) {
          status = "skipped";
        } else {
          status = "failed";
          pipelineFailed = true;
        }
      } else {
        status = "success";
        context.step_results[step.name] = result;
      }

      stepResults.push({
        name: step.name,
        status,
        started_at: stepStartedAt,
        ended_at: stepEndedAt,
        duration_seconds: stepDuration,
        result: status === "success" ? result : undefined,
        error,
        attempts,
      });

      if (pipelineFailed) {
        break;
      }
    }

    const pipelineEndedAt = new Date().toISOString();
    const pipelineDuration = (Date.now() - pipelineStartMs) / 1000;

    let stepsPassed = 0;
    let stepsFailed = 0;
    let stepsSkipped = 0;

    for (const sr of stepResults) {
      if (sr.status === "success") stepsPassed++;
      else if (sr.status === "failed") stepsFailed++;
      else if (sr.status === "skipped") stepsSkipped++;
    }

    return {
      pipeline_name: this._name,
      status: pipelineFailed ? "failed" : "success",
      started_at: pipelineStartedAt,
      ended_at: pipelineEndedAt,
      duration_seconds: pipelineDuration,
      steps: stepResults,
      steps_passed: stepsPassed,
      steps_failed: stepsFailed,
      steps_skipped: stepsSkipped,
    };
  }
}

/**
 * Format a pipeline result as a human-readable string.
 * @param {object} result - PipelineResult from pipeline.run()
 * @returns {string}
 */
export function formatResult(result) {
  const lines = [];
  const total = result.steps.length;

  lines.push(`=== Pipeline: ${result.pipeline_name} ===`);
  lines.push(`Status: ${result.status}`);
  lines.push(`Duration: ${result.duration_seconds.toFixed(2)}s`);
  lines.push(
    `Steps: ${total} total | ${result.steps_passed} passed | ${result.steps_failed} failed | ${result.steps_skipped} skipped`,
  );
  lines.push("");

  for (const step of result.steps) {
    let tag;
    if (step.status === "success") tag = "PASS";
    else if (step.status === "skipped") tag = "SKIP";
    else tag = "FAIL";

    let detail = `${step.duration_seconds.toFixed(2)}s`;
    if (step.attempts > 1) {
      detail += `, ${step.attempts} attempts`;
    }

    let line = `  [${tag}] ${step.name} (${detail})`;
    if (step.error !== null) {
      line += ` — ${step.error}`;
    }
    lines.push(line);
  }

  return lines.join("\n");
}
